platform/fabric/events_test.go

```diff
@@ -18,74 +17,88 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-type mockSubscriber struct {
-	listener events.Listener
-	m        sync.RWMutex
-}
-
-func (m *mockSubscriber) Subscribe(chaincodeName string, listener events.Listener) {
-	m.m.Lock()
-	defer m.m.Unlock()
-	m.listener = listener
-}
-
-func (m *mockSubscriber) Unsubscribe(chaincodeName string, listener events.Listener) {
-	m.m.Lock()
-	defer m.m.Unlock()
-	m.listener = nil
-}
-
-func (m *mockSubscriber) Publish(chaincodeName string, event *committer.ChaincodeEvent) {
-	m.m.RLock()
-	defer m.m.RUnlock()
-	if m.listener != nil {
-		m.listener.OnReceive(event)
-	}
-}
+const (
+	timeout = 75 * time.Millisecond
+	waitFor = 4 * timeout
+	tick    = timeout / 10
+)
 
 func TestEventListener(t *testing.T) {
 	subscriber := &mockSubscriber{}
 	listener := newEventListener(subscriber, "testChaincode")
-	eventChannel := listener.ChaincodeEvents()
+	ch := listener.ChaincodeEvents()
+
+	done := make(chan bool)
 
 	var wg sync.WaitGroup
+	wg.Add(2)
 
 	// Publish events
-	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		for i := 0; i < 1000; i++ {
-			subscriber.Publish("testChaincode", &committer.ChaincodeEvent{Payload: []byte(fmt.Sprintf("event %d", i))})
-			time.Sleep(1 * time.Millisecond)
+		msg := []byte("some msg")
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				subscriber.Publish("testChaincode", &committer.ChaincodeEvent{Payload: msg})
+			}
 		}
 	}()
 
-	wg.Add(1)
 	// Stop while the publisher is still publishing
-	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
+
 	go func() {
 		defer wg.Done()
 		for {
 			select {
-			case event := <-eventChannel:
+			case event := <-ch:
+				// we got a new event
 				assert.NotNil(t, event)
 			case <-ctx.Done():
+				// our timeout is fired
+				// this should close our channel
 				listener.CloseChaincodeEvents()
 				return
 			}
 		}
 	}()
+
+	assert.Eventually(t, func() bool {
+		// eventually our event should be closed
+		_, ok := <-ch
+		return !ok
+	}, waitFor, tick)
+
+	// now we let our consumer know that they can stop working
+	close(done)
 	wg.Wait()
-	time.Sleep(100 * time.Millisecond)
+}
 
-	// Ensure the channel is closed
-	select {
-	case _, ok := <-eventChannel:
-		if ok {
-			t.Fatal("expected channel to be closed")
-		}
-	default:
-		t.Fatal("expected to read from closed channel")
+type mockSubscriber struct {
```

Comment:
(super minor nit, more like a philosophical comment): it's always useful to minimise diffs if possible so that reviewers can see what really changed. The diff chunk below exists but nothing actually changed. It would have been more efficient to actually leave the code where it was.

Reply:
yeah - I agree